src/shared/test/integration/inter-frame-communication-test.js

```diff
@@ -0,0 +1,243 @@
+import { delay } from '../../../test-util/wait';
+import { Bridge as BridgeOri } from '../../bridge';
+import { ListenerCollection as ListenerCollectionOri } from '../../listener-collection';
+import { PortFinder as PortFinderOri } from '../../port-finder';
+import { PortProvider as PortProviderOri } from '../../port-provider';
+import { isMessageEqual } from '../../port-util';
+
+describe('PortProvider-PortFinder-Bridge integration', () => {
+  let destroyables;
+
+  // These proxy classes are to keep track of instances that need to be destroyed.
+  class Bridge extends BridgeOri {
+    constructor() {
+      super();
+      destroyables.push(this);
+    }
+  }
+
+  class ListenerCollection extends ListenerCollectionOri {
+    constructor() {
+      super();
+      destroyables.push(this);
+    }
+    destroy() {
+      this.removeAll();
+    }
+  }
+
+  class PortFinder extends PortFinderOri {
+    constructor(...args) {
+      super(...args);
+      destroyables.push(this);
+    }
+  }
+
+  class PortProvider extends PortProviderOri {
+    constructor(...args) {
+      super(...args);
+      destroyables.push(this);
+    }
+  }
+
+  beforeEach(() => {
+    destroyables = [];
+  });
+
+  afterEach(() => {
+    destroyables.forEach(instance => instance.destroy());
+  });
+
+  it('enables the communication between sidebar-host and notebook-sidebar frames', async () => {
+    let done;
+    const promise = new Promise(resolve => (done = resolve));
+
+    // Test this fictional message path:
+    // host -- ping --> sidebar -- ping --> notebook
+    // host <- pong --- sidebar <- pong --- notebook
+
+    // host frame
+    const portProvider = new PortProvider(window.location.origin);
+    const hostBridge = new Bridge();
+
+    // register RPC method *before* creating the channel
+    hostBridge.on('pong', done);
+    hostBridge.createChannel(portProvider.sidebarPort);
+
+    // call RPC method *after* creating the channel. These RPC messages are
+    // queued and will be eventually be received by the sidebar.
+    hostBridge.call('ping');
+
+    portProvider.listen();
+
+    await delay(10); // add some realism
+
+    // sidebar frame
+    const portFinder1 = new PortFinder({
+      hostFrame: window,
+      source: 'sidebar',
+    });
+    const sidebarBridge1 = new Bridge();
+    const sidebarBridge2 = new Bridge();
+
+    // The following promise is to prevent sending a message to the notebook
+    // before the notebook-sidebar (sidebarBridge2) channel is created.
+    let notebookSidebarEnable;
+    const isNotebookSidebarEnable = new Promise(
+      resolve => (notebookSidebarEnable = resolve)
+    );
+
+    // register RPC method *before* creating the channel.
+    sidebarBridge1.on('ping', async () => {
+      await isNotebookSidebarEnable;
+      sidebarBridge2.call('ping');
+    });
+    sidebarBridge2.on('pong', () => sidebarBridge1.call('pong'));
```

Comment:
One way to avoid to instantiate two `Bridge`s, could be to allow the bridge to forward messages to specific frames, in this way:
```suggestion
    const sidebarBridge = new Bridge();

    // The following promise is to prevent sending a message to the notebook
    // before the notebook-sidebar (sidebarBridge2) channel is created.
    let notebookSidebarEnable;
    const isNotebookSidebarEnable = new Promise(
      resolve => (notebookSidebarEnable = resolve)
    );

    // register RPC method *before* creating the channel.
    sidebarBridge.on('ping', async () => {
      await isNotebookSidebarEnable;
      sidebarBridge.call('ping', {to: 'notebook'});
    });
    sidebarBridge2.on('pong', () => sidebarBridge.call('pong', {to: 'host'}));
```